lib/galaxy/tools/parameters/basic.py

```diff
@@ -2177,12 +2184,12 @@ def from_json(self, value, trans, other_values=None):
                         v = v.hda
                     match = dataset_matcher.hda_match(v)
                     if match and match.implicit_conversion:
-                        v.implicit_conversion = True
+                        v.implicit_conversion = True  # type:ignore[union-attr]
         if not self.multiple:
             if len(rval) > 1:
                 raise ParameterValueError("more than one dataset supplied to single input dataset parameter", self.name)
             if len(rval) > 0:
-                rval = rval[0]
+                rval = rval[0]  # type:ignore[assignment]
```

Comment:
```suggestion
                return rval[0]
```
probably better than ignoring the type